### PR TITLE
[WIP] Add nested get query parameter expandation

### PR DESCRIPTION
**Pull request by iguchi1124**
I'm creating this issue https://github.com/sinatra/mustermann/issues/61 examples in this pull request!
What do you think about this approach?

```rb
pattern = Mustermann.new("/books/:id")
pattern.expand(:append, id: 5, include: ["foo"]) # => "/books/5?include[]=foo"
pattern.expand(:append, id: 5, foo: {bar: :baz}) # => "/books/5?foo[bar]=baz"
pattern.expand(:append, id: 5, foo: {bar: [1, 2]}) # => "/books/5?foo[bar][]=1&foo[bar][]=2"

```

### CODE REVIEW

File: mustermann/mustermann.gemspec [hunk 1]
```diff
@@ -14,4 +14,6 @@ Gem::Specification.new do |s|
   s.files                 = `git ls-files`.split("\n")
   s.test_files            = `git ls-files -- {test,spec,features}/*`.split("\n")
   s.executables           = `git ls-files -- bin/*`.split("\n").map{ |f| File.basename(f) }
+
+  s.add_dependency 'query_string_builder'
```
Comment (iguchi1124) [inline mustermann/mustermann.gemspec]: I can remove this dependency, it is my sample code

**Comment by iguchi1124** [comment]
inlin-ized and fixup this commit https://github.com/sinatra/mustermann/pull/84/commits/1e5a12d25c842c217a65acf6a6a6de8608c0e4da

File: mustermann/lib/mustermann/expander.rb [hunk 2]
```diff
@@ -194,8 +194,7 @@ def slice(hash, keys)
 
     def append(uri, values)
       return uri unless values and values.any?
-      entries = values.map { |pair| pair.map { |e| @api_expander.escape(e, also_escape: /[\/\?#\&\=%]/) }.join(?=) }
-      "#{ uri }#{ uri[??]??&:?? }#{ entries.join(?&) }"
+      [uri, QueryStringBuilder.new(values).build].join('?')
```
Comment (namusyaka) [inline mustermann/lib/mustermann/expander.rb]: I don't think `QueryStringBuilder` is necessary.  Recursive processing looks fine, but `QueryStringBuilder` is really overkill.

File: mustermann/lib/mustermann/expander.rb [hunk 3]
```diff
@@ -194,8 +194,7 @@ def slice(hash, keys)
 
     def append(uri, values)
       return uri unless values and values.any?
-      entries = values.map { |pair| pair.map { |e| @api_expander.escape(e, also_escape: /[\/\?#\&\=%]/) }.join(?=) }
-      "#{ uri }#{ uri[??]??&:?? }#{ entries.join(?&) }"
```
Comment (namusyaka) [inline mustermann/lib/mustermann/expander.rb]: Why did you delete these lines? Your patch doesn't cover these feature.

File: mustermann/spec/expander_spec.rb [hunk 4]
```diff
@@ -70,6 +70,9 @@
       subject(:expander) { Mustermann::Expander.new('/:a', additional_values: :append) }
       example { expander.expand(a: ?a).should be == '/a' }
       example { expander.expand(a: ?a, b: ?b).should be == '/a?b=b' }
+      example { expander.expand(a: ?a, b: [?b]).should be == '/a?b[]=b' }
+      example { expander.expand(a: ?a, b: {c: ?c}).should be == '/a?b[c]=c' }
+      example { expander.expand(a: ?a, b: {c: [1, 2]}).should be == '/a?b[c][]=1&b[c][]=2' }
```
Comment (namusyaka) [inline mustermann/spec/expander_spec.rb]: The original test doesn't mention escape, but you need to add about it at least to add this change.

**Comment by iguchi1124** [comment]
Thank you for reviews! I'll update this pull request soon.

**Comment by namusyaka** [comment]
@iguchi1124 If you're going to include this fix in the 1.0.3 release, please write to that effect in [the milestone issue](https://github.com/sinatra/mustermann/issues/87).

File: mustermann/lib/mustermann/expander.rb [hunk 5]
```diff
@@ -2,6 +2,7 @@
 require 'mustermann/ast/expander'
 require 'mustermann/caster'
 require 'mustermann'
+require 'addressable/uri'
```
Comment (namusyaka) [inline mustermann/lib/mustermann/expander.rb]: What is this?
I would not like to use the `addressable` gem. Mustermann has not been depended on external gems.
Is there a clear reason to add this gem?
If so, why do not you add this to gemspec.

**Comment by namusyaka** [comment]
@iguchi1124 Hey, I'm going to release new version next week. So if you'd like to include this fix at the next version, please ping me.
